modules/efk-logging-fluentd-undefined.adoc

```diff
@@ -0,0 +1,121 @@
+top-level fields. You can add undefined fields to the top-level fields and move others to an `undefined` container.
++
+You can also convert undefined fields to JSON format and replace special characters.
+
+* Defined fields. You can also configure which defined fields appear in the top levels of the logs. 
++
+[[default-fields]]
+The default top-level fields are `CEE,time`, `@timestamp`, `aushape`, `ci_job`, `collectd`, `docker`, `fedora-ci`, 
+`file`, `foreman`, `geoip`, `hostname`, `ipaddr4`, `ipaddr6`, `kubernetes`, `level`, `message`, `namespace_name`, `namespace_uuid`,
+`offset`, `openstack`, `ovirt`, `pid`, `pipeline_metadata`, `rsyslog`, `service`, `systemd`, `tags`, `testcase`, `tlog`, `viaq_msg_id`.
+You can use the `CDM_DEFAULT_KEEP_FIELDS` parameter to add additional defaults and the `CDM_EXTRA_KEEP_FIELDS` parameter to include 
+additional non-default fields at the top level.
++
+Any defined fields not included in `${CDM_DEFAULT_KEEP_FIELDS}` or `${CDM_EXTRA_KEEP_FIELDS}` are moved to `${CDM_UNDEFINED_NAME}` if `CDM_USE_UNDEFINED` is `true`.
+
+* Empty fields. You can determine whether to retain empty fields from disparate logs.
+
+.Environment parameters for log normalization
+[cols="2,4,2", options="header"]
+|===
+|Parameters  |Definition  |Example
+
+|CDM_DEFAULT_KEEP_FIELDS
+|Specify a set of defined fields add as default top-level fields. 
+|See the xref:default-fields[list of defaults].
+	-
+|CDM_EXTRA_KEEP_FIELDS
+|Specify an extra set of defined fields to be kept in addition to the `CDM_DEFAULT_KEEP_FIELDS`. The default is "".
+|CDM_EXTRA_KEEP_FIELDS="broker"
+
+|CDM_KEEP_EMPTY_FIELDS
+|Specify defined fields to retain even if empty. Empty defined fields not specified are dropped. The default is "".
+|CDM_KEEP_EMPTY_FIELDS="offset"
+
+|CDM_USE_UNDEFINED
+|Set to `true` to move undefined fields to the `undefined` top level field. The default is `false`. If `true`, values in `CDM_DEFAULT_KEEP_FIELDS`
+and `CDM_EXTRA_KEEP_FIELDS` are not moved to `undefined`.
+|CDM_USE_UNDEFINED=true
+
+|CDM_UNDEFINED_NAME
+|Specify a name for the undefined top level field if using `CDM_USE_UNDEFINED`. The default is`undefined`.
+Enabled only when `CDM_USE_UNDEFINED` is `true`.
+|CDM_UNDEFINED_NAME="undef"
+
+|CDM_UNDEFINED_MAX_NUM_FIELDS
+|If the number of undefined fields is greater than this number, all undefined fields are converted to their JSON string representation 
+and stored in the `CDM_UNDEFINED_NAME` field. If the record contains more than this value of undefined fields, no further processing takes place on these fields. 
+Instead, the fields will be converted to a single string JSON value, stored in the top-level `CDM_UNDEFINED_NAME` field. Keeping the default of `-1`allows for an unlimited number of undefined fields, which is not recommended.
+Enabled only when `CDM_USE_UNDEFINED` is `true`.
```

Comment:
According to https://jira.coreos.com/browse/LOG-372?focusedCommentId=114310&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-114310, thought this line should be removed.

Reply:
Sorry. I missed Rich's message on this from yesterday.

Reply:
Rather, can we add a note that this [1] is executed even if `CDM_USE_UNDEFINED` is false?  That'd reduce the confusion (as I had... :)

[1]  "the fields will be converted to a single string JSON value, stored in the top-level `CDM_UNDEFINED_NAME` field"

Reply:
@nhosoi Added a note. Does this meet your expectations? 
5af4d98f06a472449bbc865b3586e2f55278dcd9